upgrade/upgrade_common_services.sh

```diff
@@ -202,34 +186,109 @@ function compare_channel() {
+            success "current channel ${cur_channel} is less than upgrade channel ${channel}, ready for channel switch"
             return 0
         fi
     done
-    success "Upgrade channel ${channel} is equal to current channel ${cur_channel}, ready for channel switch."
+    CHANNEL_COMP=0
+    success "current channel ${cur_channel} is equal to upgrade channel ${channel}, do not need channel switch"  
+}
+
+function deployment_check(){
+    local subName=$1
+    local csNS=$2
+    local channel=$3
+
+    STEP=$((STEP + 1 ))
+    msg ""
+    title "[${STEP}] Checking ${subName} deployment in ${csNS} namespace..."
+    msg "-----------------------------------------------------------------------"
+
+    # get current cs opertor channel version 
+    csoperator_channel=$(oc get sub -n ${csNS} | grep ${subName} | awk '{print $4}')
+    compare_channel "${subName}" "${csNS}" "${channel}" "${csoperator_channel}"
+    
+
+    if [[ $CHANNEL_COMP == 1 ]]; then
+        msg "current channel version of ${subName} ${csoperator_channel} is less then upgrade channel version ${channel}"
+        msg ""
+
+        in_step=1
+        # scale down cs operator to prevent reconciliation
+        msg "[${in_step}] Scaling down ${subName} deployment in ${csNS} namespace to 0"
+        oc scale deployment -n "${csNS}" "${subName}" --replicas=0
+
+        # delete OperandRegistry
+        in_step=$((in_step + 1))
+        msg "[${in_step}] Deleting OperandRegistry common-service in ${csNS} namespace..."
+        oc delete opreg common-service -n ${csNS} --ignore-not-found
+        
+    elif [[ $CHANNEL_COMP != 1 ]]; then
+        msg "current channel version of ${subName} ${csoperator_channel} is not less than upgrade channel version ${channel}"
+        msg ""
+
+        # get installedCSV from subscription
+        csv=$(oc get sub ${subName} -n ${csNS} -o=jsonpath='{.status.installedCSV}' --ignore-not-found)
+        msg "existing installedCSV is ${csv}"
+
+        # remove all chars before "v"
+        trimmed_csv="$(echo $csv | awk -Fv '{print $NF}')"
+        trimmed_channel="$(echo $channel | awk -Fv '{print $NF}')"
+
+        if [[ "$trimmed_csv" == *"$trimmed_channel"* ]]; then
+            in_step=1
+            msg "installedCSV ${csv} matches upgrade channel version ${channel}"
+            msg ""
+            # scale up cs operator back to 1
+            msg "[${in_step}] Scaling up ${subName} deployment in ${csNS} namespace to 1"
+            oc scale deployment -n "${csNS}" "${subName}" --replicas=1
+        fi
+    fi
 }
 
+function pre-zen(){
+    local csNS=$1
+
+    STEP=$((STEP + 1 ))
+    msg ""
+    title "[${STEP}] Detecting Operator Condition of zen operator v1.4.2 in ${csNS} namespace..."
+    msg "-----------------------------------------------------------------------"
+
+    if oc get operatorcondition -n ${csNS} 2>/dev/null | grep ibm-zen-operator ; then
+        msg "Removing Operator Condition of zen operator v1.4.2..."
```

Comment:
Why would they remove the operator condition? Isn't this in place by the operator to protect against this sort of interruption? Has the Zen team signed off here?

Reply:
Brought this up in the roadmap issue to Yuchen Shen: https://github.ibm.com/IBMPrivateCloud/roadmap/issues/55451#issuecomment-50065505

Reply:
it was discussed and agreed with the Zen team, to address the kind of race conditional which would happen on upgrade from old version of Common Services / Zen into the latest.

Reply:
From Yuchen and Piotr: https://github.ibm.com/PrivateCloud-analytics/CPD-Quality/issues/6163#issuecomment-49840184

> ### Immediate workarounds
> Remove operatorcondition for ibm-zen-operator 1.4.2
> 
> ➜  ~ oc get operatorcondition -n ibm-common-services
> NAME                                AGE
> ibm-cert-manager-operator.v3.24.0   7d1h
> ibm-zen-operator.v1.4.2             7d1h
> ibm-zen-operator.v1.8.0             7d1h
> ➜  ~ oc delete operatorcondition ibm-zen-operator.v1.4.2 -n ibm-common-services
> operatorcondition.operators.coreos.com "ibm-zen-operator.v1.4.2" deleted
> ### Action
> After interlock meeting with Zen team, we are on the same page that pre-zen job could be run before the zen upgrade happens, and it will help to unblock subsequent upgrade.
> 
> ### New Remediation
> We will integrate pre-zen job into upgrade script, it will help to delete operatorcondition for ibm-zen-operator 1.4.2 if it does exist.
> Deletion of ibm-zen-operator 1.4.2 OperandCondition will not be executed if it is an upgrade for other version of zen.
> 
> We expect to deliver the script with pre-zen job integration next Monday, and olm-utils will pick it up during the image build.

Reply:
@taylormgeorge91 -- from Yuchen and Piotr: https://github.ibm.com/PrivateCloud-analytics/CPD-Quality/issues/6163#issuecomment-49840184

> ### Immediate workarounds
> Remove operatorcondition for ibm-zen-operator 1.4.2
> 
> ➜  ~ oc get operatorcondition -n ibm-common-services
> NAME                                AGE
> ibm-cert-manager-operator.v3.24.0   7d1h
> ibm-zen-operator.v1.4.2             7d1h
> ibm-zen-operator.v1.8.0             7d1h
> ➜  ~ oc delete operatorcondition ibm-zen-operator.v1.4.2 -n ibm-common-services
> operatorcondition.operators.coreos.com "ibm-zen-operator.v1.4.2" deleted
> ### Action
> After interlock meeting with Zen team, we are on the same page that pre-zen job could be run before the zen upgrade happens, and it will help to unblock subsequent upgrade.
> 
> ### New Remediation
> We will integrate pre-zen job into upgrade script, it will help to delete operatorcondition for ibm-zen-operator 1.4.2 if it does exist.
> Deletion of ibm-zen-operator 1.4.2 OperandCondition will not be executed if it is an upgrade for other version of zen.
> 
> We expect to deliver the script with pre-zen job integration next Monday, and olm-utils will pick it up during the image build.